linux/src/socket_transport_linux.c

```diff
@@ -0,0 +1,696 @@
+    else
+    {
+        if ((opts = fcntl(socket, F_SETFL, opts|O_NONBLOCK)) < 0)
+        {
+            LogErrorNo("Failure setting socket option.");
+            result = MU_FAILURE;
+        }
+        else
+        {
+            result = 0;
+        }
+    }
+    return result;
+}
+
+static SOCKET_HANDLE connect_to_client(const char* hostname, uint16_t port, uint32_t connection_timeout)
+{
+    SOCKET_HANDLE result;
+    // Codes_SOCKET_TRANSPORT_LINUX_11_015: [ socket_transport_connect shall call socket with the params AF_INET, SOCK_STREAM and 0. ]
+    SOCKET_HANDLE client_socket = socket(AF_INET, SOCK_STREAM, 0);
+    if (client_socket == INVALID_SOCKET)
+    {
+        LogErrorNo("Failure: socket create failure.");
+        result = INVALID_SOCKET;
+    }
+    else
+    {
+        char port_string[16];
+        if (sprintf(port_string, "%u", port) < 0)
+        {
+            LogError("failed to copy port value %" PRIu16 "", port);
+            result = INVALID_SOCKET;
+        }
+        else
+        {
+            struct addrinfo addr_hint = { 0 };
+            struct addrinfo* addrInfo = NULL;
+
+            addr_hint.ai_family = AF_INET;
+            addr_hint.ai_socktype = SOCK_STREAM;
+            addr_hint.ai_protocol = 0;
+            addr_hint.ai_flags = AI_CANONNAME;
+
+            if (getaddrinfo(hostname, port_string, &addr_hint, &addrInfo) != 0)
+            {
+                LogErrorNo("Failure: getaddrinfo(hostname=%s, portString=%s, &addr_hint=%p, &addrInfo=%p)", hostname, port_string, &addr_hint, &addrInfo);
+                result = INVALID_SOCKET;
+            }
+            else
+            {
+                LogInfo("Connecting to %s:%" PRIu16 " Machine Name: %s, connection timeout: %" PRIu32 "", hostname, port, MU_P_OR_NULL(addrInfo->ai_canonname), connection_timeout);
+
+                // Codes_SOCKET_TRANSPORT_LINUX_11_016: [ socket_transport_connect shall call connect to connect to the endpoint. ]
+                if (connect(client_socket, addrInfo->ai_addr, addrInfo->ai_addrlen) != 0)
+                {
+                    LogErrorNo("Connection failure. Server: %s:%" PRIu16 ".", hostname, port);
+                    result = INVALID_SOCKET;
+                }
+                // Codes_SOCKET_TRANSPORT_LINUX_11_017: [ socket_transport_connect shall set the socket to non-blocking by calling fcntl with O_NONBLOCK. ]
+                else if (set_nonblocking(client_socket) != 0)
+                {
+                    LogError("Failure: nonblocking failure.");
+                    result = INVALID_SOCKET;
+                }
+                else
+                {
+                    freeaddrinfo(addrInfo);
+                    result = client_socket;
+                    goto all_ok;
+                }
+                freeaddrinfo(addrInfo);
+            }
+        }
+        close(client_socket);
+    }
+all_ok:
+    return result;
+}
+
+SOCKET_TRANSPORT_HANDLE socket_transport_create(SOCKET_TYPE type)
```

Comment:
Generally having 2 APIs socket_transport_create_server and socket_transport_create_client is better, because the check is done at compile time. As in one cannot compile some code with some weird type passed in.
The only case where such a construct would be beneficial is for interfaces where there will be lots of different values added (like a set_option API).
But in all other cases, which is most, it's better to have separate APIs which make it explicitly checked at compile time.

Code will work with this too, but I'd say strong consider.

Reply:
https://msazure.visualstudio.com/One/_sprints/taskboard/Azure%20Messaging%20Store/One/Custom/AzureMessaging/Dilithium/Dt2?workitem=28591226